File: backup/cyclesignal_flutter/lib/data/synthetic_data.dart
import 'package:uuid/uuid.dart';
import '../models/cycle_entry.dart';
import '../logic/cycle.dart';

// Demo user "Amara" — PRD Section 8. Fake data only (the AI layers are real).
// 6 cycles, lengths 42/38/55/61/44/39 (irregular). Premenstrual mood crashes (sev 4-5),
// functional impact in 4 of 6, acne in 3 of 6, fatigue in 5 of 6. No safety-trigger text.

const _uuid = Uuid();
const _lengths = {1: 42, 2: 38, 3: 55, 4: 61, 5: 44, 6: 39};
const _functionalImpactCycles = {1, 2, 3, 4};
const _acneCycles = {2, 4, 6};
const _fatigueCycles = {1, 2, 3, 4, 5};

CycleEntry _empty(String date, int day, int n, String consent) => CycleEntry(
      id: _uuid.v4(),
      date: date,
      cycleDay: day,
      cycleNumber: n,
      physical: PhysicalEntry(),
      bleeding: BleedingEntry(),
      functionalImpact: FunctionalImpact(),
      source: 'synthetic',
      consentTimestamp: consent,
    );

List<CycleEntry> generateSyntheticEntries(String consent) {
  final today = todayDate();
  final todayIso = toIsoDate(today);

  final starts = <int, DateTime>{};
  starts[6] = addDays(today, -23); // current cycle on day 24 (PRD: "you're on day 24")
  for (var n = 5; n >= 1; n--) {
    starts[n] = addDays(starts[n + 1]!, -_lengths[n]!);
  }

  final entries = <CycleEntry>[];

  for (var n = 1; n <= 6; n++) {
    final L = _lengths[n]!;
    final start = starts[n]!;
    String at(int day) => toIsoDate(addDays(start, day - 1));

    // Menstruation days 1-4
    final bleed = [
      [1, 'medium', 1, 0],
      [2, 'medium', 0, 0],
      [3, 'light', 0, 0],
      [4, 'light', 0, 1],
    ];
    for (final b in bleed) {
      final e = _empty(at(b[0] as int), b[0] as int, n, consent);
      entries.add(CycleEntry(
        id: e.id,
        date: e.date,
        cycleDay: e.cycleDay,
        cycleNumber: e.cycleNumber,
        physical: PhysicalEntry(),
        bleeding: BleedingEntry(
            started: b[2] == 1, ended: b[3] == 1, flow: b[1] as String),
        functionalImpact: FunctionalImpact(),
        source: 'synthetic',
        consentTimestamp: consent,
      ));
    }

    // Acne mid-cycle (day 10) for selected cycles
    if (_acneCycles.contains(n)) {
      final e = _empty(at(10), 10, n, consent);
      entries.add(CycleEntry(
        id: e.id,
        date: e.date,
        cycleDay: 10,
        cycleNumber: n,
        physical: PhysicalEntry(acne: true),
        bleeding: BleedingEntry(),
        functionalImpact: FunctionalImpact(),
        source: 'synthetic',
        consentTimestamp: consent,
      ));
    }

    // Premenstrual mood crashes — days L-6, L-4, L-2
    final premen = [
      {'day': L - 6, 'sev': 4, 'types': ['low'], 'sudden': false, 'idx': 0},
      {'day': L - 4, 'sev': 5, 'types': ['low', 'tearful'], 'sudden': true, 'idx': 1},
      {'day': L - 2, 'sev': 4, 'types': ['low', 'irritable'], 'sudden': false, 'idx': 2},
    ];
    for (final p in premen) {
      final day = p['day'] as int;
      final idx = p['idx'] as int;
      Severity? fatigue;
      SleepEntry? sleep;
      if (_fatigueCycles.contains(n) && idx == 1) {
        fatigue = Severity(present: true, severity: 4);
        sleep = SleepEntry(disrupted: true, nightsAffected: 3);
      }
      FunctionalImpact fi = FunctionalImpact();
      if (_functionalImpactCycles.contains(n) && idx == 2) {
        fi = FunctionalImpact(
          present: true,
          types: [n.isEven ? 'cancelled_plans' : 'missed_work'],
          severity: 'moderate',
        );
      }
      entries.add(CycleEntry(
        id: _uuid.v4(),
        date: at(day),
        cycleDay: day,
        cycleNumber: n,
        mood: MoodEntry(
          present: true,
          types: (p['types'] as List).cast<String>(),
          severity: p['sev'] as int,
          suddenOnset: p['sudden'] as bool,
        ),
        physical: PhysicalEntry(fatigue: fatigue, sleep: sleep),
        bleeding: BleedingEntry(),
        functionalImpact: fi,
        source: 'synthetic',
        consentTimestamp: consent,
      ));
    }
  }

  // Never seed entries in the future (current cycle is only partway through).
  final filtered = entries.where((e) => e.date.compareTo(todayIso) <= 0).toList()
    ..sort((a, b) => a.date.compareTo(b.date));
  return filtered;
}

File: backup/cyclesignal_rn/app/src/data/syntheticData.ts
import { CycleEntry, FunctionalImpactType, MoodType, Severity } from '../types';
import { createId } from '../lib/id';
import { addDays, toISODate } from '../lib/cycle';

// Demo user "Amara" — PRD Section 8.
// 6 cycles, lengths 42/38/55/61/44/39 (irregular, PCOS-consistent). Premenstrual mood
// crashes (sev 4-5), functional impact in 4 of 6, acne in 3 of 6, fatigue in 5 of 6.
// No safety-trigger language. All entries source: "synthetic".

const LENGTHS: Record<number, number> = { 1: 42, 2: 38, 3: 55, 4: 61, 5: 44, 6: 39 };
const FUNCTIONAL_IMPACT_CYCLES = new Set([1, 2, 3, 4]); // 4 of 6
const ACNE_CYCLES = new Set([2, 4, 6]); // 3 of 6
const FATIGUE_CYCLES = new Set([1, 2, 3, 4, 5]); // 5 of 6
const IMPACT_TYPES: FunctionalImpactType[][] = [['missed_work'], ['cancelled_plans']];

function emptyEntry(
  date: string,
  cycleDay: number,
  cycleNumber: number,
  consentTimestamp: string
): CycleEntry {
  return {
    id: createId(),
    date,
    cycleDay,
    cycleNumber,
    mood: null,
    physical: { pain: null, acne: false, hairGrowth: false, fatigue: null, sleep: null, bloating: false, other: [] },
    bleeding: { started: false, ended: false, flow: null },
    functionalImpact: { present: false, types: [], severity: null },
    medications: [],
    emotionSnapshot: null,
    conversationTranscript: '',
    source: 'synthetic',
    safetyFlagged: false,
    consentTimestamp,
  };
}

export function generateSyntheticEntries(consentTimestamp: string): CycleEntry[] {
  const today = new Date(new Date().toISOString().slice(0, 10) + 'T00:00:00');
  const todayISO = toISODate(today);

  // Anchor: current cycle (6) is on day 24 today (PRD: "You're on day 24").
  const starts: Record<number, Date> = {};
  starts[6] = addDays(today, -23);
  for (let n = 5; n >= 1; n--) starts[n] = addDays(starts[n + 1], -LENGTHS[n]);

  const entries: CycleEntry[] = [];

  for (let n = 1; n <= 6; n++) {
    const L = LENGTHS[n];
    const start = starts[n];
    const at = (day: number) => toISODate(addDays(start, day - 1));

    // Menstruation, days 1-4
    const bleed: Array<[number, 'medium' | 'light', boolean, boolean]> = [
      [1, 'medium', true, false],
      [2, 'medium', false, false],
      [3, 'light', false, false],
      [4, 'light', false, true],
    ];
    for (const [day, flow, started, ended] of bleed) {
      const e = emptyEntry(at(day), day, n, consentTimestamp);
      e.bleeding = { started, ended, flow };
      entries.push(e);
    }

    // Acne mid-cycle (day 10) for selected cycles
    if (ACNE_CYCLES.has(n)) {
      const e = emptyEntry(at(10), 10, n, consentTimestamp);
      e.physical.acne = true;
      entries.push(e);
    }

    // Premenstrual mood crashes — days L-6, L-4, L-2
    const premen: Array<{ day: number; sev: Severity; types: MoodType[]; sudden: boolean }> = [
      { day: L - 6, sev: 4, types: ['low'], sudden: false },
      { day: L - 4, sev: 5, types: ['low', 'tearful'], sudden: true },
      { day: L - 2, sev: 4, types: ['low', 'irritable'], sudden: false },
    ];
    premen.forEach((p, idx) => {
      const e = emptyEntry(at(p.day), p.day, n, consentTimestamp);
      e.mood = { present: true, types: p.types, severity: p.sev, suddenOnset: p.sudden };
      if (FATIGUE_CYCLES.has(n) && idx === 1) {
        e.physical.fatigue = { present: true, severity: 4 };
        e.physical.sleep = { disrupted: true, nightsAffected: 3 };
      }
      if (FUNCTIONAL_IMPACT_CYCLES.has(n) && idx === 2) {
        e.functionalImpact = {
          present: true,
          types: IMPACT_TYPES[n % IMPACT_TYPES.length],
          severity: 'moderate',
        };
      }
      entries.push(e);
    });
  }

  // Never seed entries in the future (the current cycle is only partway through).
  return entries
    .filter((e) => e.date <= todayISO)
    .sort((a, b) => a.date.localeCompare(b.date));
}

// The cycle/day a fresh voice check-in should attach to (current cycle, today).
export function currentCyclePosition(): { cycleNumber: number; cycleDay: number } {
  return { cycleNumber: 6, cycleDay: 24 };
}
